### PR TITLE
doc: subsystem: Settings subsystem documentation

### PR DESCRIPTION
**Pull request by nvlsianpu**
This patch introduce doc for settings serialization
subsystem with FCB and File System beck-ends.

This is addendum to the #6408 PR

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

**Comment by carlescufi** [comment]
@nvlsianpu shouldn't we first decide on the prefix we are going to use in the code itself? or is `conf_` already decided for that?

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 1]
```diff
@@ -0,0 +1,156 @@
+.. _conf_s11n
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: add a colon

    .. _conf_s11n:

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+Runtime configuration system
+############################
+
+Config subsystem is meant for persisting per-device configuration
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: The config subsystem gives modules a way to store persistent
    per-device configuration and runtime state.

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+and runtime state for modules.
+
+Configuration items are stored as key-value pairs, where both the key and
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: How about:

    Configuration items are stored as key-value pair strings.  By convention,
    the keys can be organized by the package and subtree defining the key,
    for example the key ``id/serial`` would define the ``serial`` configuration
    element for the package ``id``.

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+the value are expected to be strings. Keys are divided into component
+elements, where packages register their subtree using the first element.
+E.g key ``id/serial`` consists of 2 components, ``id`` and ``serial``.
+Package sys/id registered it's subtree of configuration elements to be
+under ``id``.
+
+There are convenience routines for converting value back and forth
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: How about:

    Convenience routines are provided for converting a key value to
    and from a string type.

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        k_sleep(1000);
+        sys_reboot(SYS_REBOOT_COLD);
+    }
+
+API
+***
+
+The API of runtime configuration APIs are provided by conf_s11n.h:
+
+.. doxygengroup:: sys_config
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: #6408 needs to be merged before this reference will not generate an error.

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+from string.
+
+Handlers
+********
+
+Config handlers for subtree implement a set of handler functions.
+These are registered using a call to ``conf_register()``.
+
+**ch_get**
+    This gets called when somebody asks for a config element value
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: This gets called when asking for a config element value by its name
    using ``conf_get_value()``.

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    by it's name using ``conf_get_value()``.
+
+**ch_set**
+    This is called when value is being set using ``conf_set_value()``, and
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: This gets called when the value ...

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    also when configuration is loaded from persisted storage with
+    ``conf_load()``.
+
+**ch_commit**
+    This gets called after configuration has been loaded in full.
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: after the configuration

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+Config handlers for subtree implement a set of handler functions.
+These are registered using a call to ``conf_register()``.
+
+**ch_get**
+    This gets called when somebody asks for a config element value
+    by it's name using ``conf_get_value()``.
+
+**ch_set**
+    This is called when value is being set using ``conf_set_value()``, and
+    also when configuration is loaded from persisted storage with
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: when the configuration

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    ``conf_load()``.
+
+**ch_commit**
+    This gets called after configuration has been loaded in full.
+    Sometimes you don't want individual configuration value to take
+    effect right away, for example if there are multiple settings
+    which are interdependent.
+
+**ch_export**
+    This gets called to dump all current configuration. This happens
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: Would ``write`` be better than ``dump``?

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    when ``conf_save()`` tries to save the settings or transfer to any
+    user-implemented back-end.
+
+Persistence
+***********
+
+Backend storage for the config can be either FCB, a file in filesystem,
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: in the filesystem

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+or both.
+
+You can declare multiple sources for configuration; settings from
+all of these are restored when ``conf_load()`` is called.
+
+There can be only one target for writing configuration; this is where
+data is stored when you call ``conf_save()``, or conf_save_one().
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: or ``conf_save_one()``

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+FCB read target is registered using ``conf_fcb_src()``, and write target
+using ``conf_fcb_dst()``. ``conf_fcb_src()`` as side-effect initializes the
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: As a side-effect, ``conf_fcb_src()`` initializes the ...

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+FCB area, so it must be called when calling ``conf_fcb_dst()``.
+File read target is registered using ``conf_file_src()``, and write target
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: and write target by using

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+``conf_file_dst()``.
+
+Example: Device Configuration
+*****************************
+
+This is a simple example, config handler only implements ``ch_set`` and
+``ch_export``. ``ch_set`` is called when value is restored from storage (or
+when set initially), and ``ch_export`` is used to write the value to
+storage thanks to ''storage_func'' (it is possible to implement another export
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: ``storage_func()``.  The user can also implement some other export
    functionality, for example, writing to the shell console).

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+functionality by the user e.g. dump to to shell console).
+
+.. code:: c
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: .. code-block:: c

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+        storage_func("foo/bar", buf)
+
+        return 0;
+    }
+
+Example: Persist Runtime State
+******************************
+
+This is a simple example showing how to persist runtime state. Here
+there is only ``ch_set`` defined, which is used when restoring value from
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: In this example, only ``ch_set`` is defined, ....

    restoring the value

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+FCB read target is registered using ``conf_fcb_src()``, and write target
+using ``conf_fcb_dst()``. ``conf_fcb_src()`` as side-effect initializes the
+FCB area, so it must be called when calling ``conf_fcb_dst()``.
+File read target is registered using ``conf_file_src()``, and write target
+``conf_file_dst()``.
+
+Example: Device Configuration
+*****************************
+
+This is a simple example, config handler only implements ``ch_set`` and
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: where the config handler

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+``ch_export``. ``ch_set`` is called when value is restored from storage (or
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: when the value

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+Example: Persist Runtime State
+******************************
+
+This is a simple example showing how to persist runtime state. Here
+there is only ``ch_set`` defined, which is used when restoring value from
+persisted storage.
+
+There's foo_callout function which increments foo_val, and then
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: In this example, the ``foo_callout`` function increments ``foo_val`` ...

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+persists the latest number. When system restarts, the application calls
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: When the system

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+``conf_load()`` while initializing, foo_val will continue counting up from where
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: and ``foo_val``

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+it was before restart.
+
+.. code:: c
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: .. code-block:: c

File: doc/subsystems/conf_s11n/conf_s11n.rst [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+        conf_save_one("foo/bar", bar);
+
+        k_sleep(1000);
+        sys_reboot(SYS_REBOOT_COLD);
+    }
+
+API
+***
+
+The API of runtime configuration APIs are provided by conf_s11n.h:
```
Comment (dbkinder) [inline doc/subsystems/conf_s11n/conf_s11n.rst]: The runtime configuration APIs are provided by ``conf_s11n.h``:

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Comment by nvlsianpu** [comment]
Hi all - I have updated the doc. Finally the name of this module stand as "Settings".

File: doc/subsystems/settings/settings.rst [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+    This gets called when asking for a settings element value
+    by its name using ``settings_get_value()``.
+
+**h_set**
+    This gets called when the value is being set using ``settings_set_value()``,
+    and also when setting is loaded from persisted storage with
+    ``settings_load()``.
+
+**h_commit**
+    This gets called after the settings has been loaded in full.
```
Comment (dbkinder) [inline doc/subsystems/settings/settings.rst]: ``settings have been``

File: doc/subsystems/settings/settings.rst [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+    Sometimes you don't want individual setting value to take
```
Comment (dbkinder) [inline doc/subsystems/settings/settings.rst]: ``want an individual``

File: doc/subsystems/settings/settings.rst [hunk 27]
```diff
@@ -0,0 +1,154 @@
...
+
+You can declare multiple sources for settings; settings from
+all of these are restored when ``settings_load()`` is called.
+
+There can be only one target for writing settings; this is where
+data is stored when you call ``settings_save()``, or ``settings_save_one()``.
+
+FCB read target is registered using ``settings_fcb_src()``, and write target
+using ``settings_fcb_dst()``. As a side-effect,  ``settings_fcb_src()``
+initializes the FCB area, so it must be called when calling
```
Comment (dbkinder) [inline doc/subsystems/settings/settings.rst]: ``called before calling``

**Review by dbkinder — CHANGES_REQUESTED** [review]
thanks for updating the docs.  almost there.

**Comment by nvlsianpu** [comment]
@dbkinder updated

**Review by dbkinder — APPROVED** [review]
Thanks!

**Comment by nvlsianpu** [comment]
@carlescufi can we merge this?

**Pull request merged**